engine/runtime-instrument-common/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala

```diff
@@ -221,6 +221,27 @@ class ChangesetBuilderTest
       )
     }
 
+    "multiline remove node" in {
+      val code =
+        """x ->
+          |    y = foo 5
+          |    z = foo 7
+          |    y + x""".stripMargin.linesIterator.mkString("\n")
```

Comment:
Shouldn't the `z = ...` line be follow with another _node definition_ line and not just an expression?

Reply:
> Shouldn't the z = ... line be followed with another node definition line

it doesn't matter, the line can contain just a return statement

> What exactly should the invalidated(ir, code, edit) code return when the last line doesn't have any node id?

The algorithm tracks node dependencies based on internal node ids. `invalidated` method returns the set of nodes (internal node ids) affected by the edit. In this particular test scenario, removing of the `z = foo 7` line should invalidate only the `z` node.